skyportal/facility_apis/lt.py

```diff
@@ -0,0 +1,854 @@
+
+        Parameters
+        ----------
+        request: skyportal.models.FollowupRequest
+            The request to add to the queue and the SkyPortal database.
+        """
+
+        from ..models import DBSession, FacilityTransaction
+
+        altdata = request.allocation.load_altdata()
+        if not altdata:
+            return
+
+        ltreq = IOORequest()
+        observation_payload = ltreq._build_prolog()
+        ltreq._build_project(observation_payload, request)
+        ltreq._build_inst_schedule(observation_payload, request)
+
+        headers = {
+            'Username': altdata["username"],
+            'Password': altdata["password"],
+        }
+        url = '{0}://{1}:{2}/node_agent2/node_agent?wsdl'.format(
+            'http', cfg['app.lt_host'], cfg['app.lt_port']
+        )
+        client = Client(url=url, headers=headers)
+        full_payload = etree.tostring(
+            observation_payload, encoding="unicode", pretty_print=True
+        )
+        # Send payload, and receive response string, removing the encoding tag which causes issue with lxml parsing
+        response = client.service.handle_rtml(full_payload).replace(
+            'encoding="ISO-8859-1"', ''
+        )
+        response_rtml = etree.fromstring(response)
+        mode = response_rtml.get('mode')
+
+        if mode == 'confirm':
+            request.status = 'submitted'
+        else:
+            request.status = f'rejected: {mode}'
+
+        transaction = FacilityTransaction(
+            request=http.serialize_requests_request_xml(full_payload),
+            response=http.serialize_requests_response_xml(response),
+            followup_request=request,
+            initiator_id=request.last_modified_by_id,
+        )
+
+        DBSession().add(transaction)
+
+    _observation_types = ['r', 'gr', 'gri', 'griz', 'ugriz']
+    _exposure_types = ['1x120s', '2x150s']
+
+    form_json_schema = {
+        "type": "object",
+        "properties": {
+            "observation_type": {
+                "type": "string",
+                "enum": _observation_types,
+                "default": "r",
+            },
+            "exposure_type": {
+                "type": "string",
+                "enum": _exposure_types,
+                "default": '1x120s',
+            },
+            "priority": {"type": "string", "enum": ["1", "5"], "default": "1"},
+            "LT_proposalID": {"type": "string"},
+            "start_date": {
+                "type": "string",
+                "format": "date",
+                "default": datetime.utcnow().date().isoformat(),
+                "title": "Start Date (UT)",
+            },
+            "end_date": {
+                "type": "string",
+                "format": "date",
+                "title": "End Date (UT)",
+                "default": (datetime.utcnow().date() + timedelta(days=7)).isoformat(),
+            },
+            "maximum_airmass": {
+                "title": "Maximum Airmass (1-3)",
```

Comment:
Perhaps it would be appropriate to use sliders here instead of the up/down selector? Since it seems like you want to define a specific range of values anyways. Same for "maximum_seeing" and "sky_brightness". You can use the uiSchema to specify this - should be just setting it to a range widget instead of the default updown. Not sure how much granularity you need for these fields but if you don't need it to support a huge number of significant figures, I think a slider with a step size of 0.01 or something like that would be good here.

Additionally, since you have a range of valid values, you should specify "minimum" and "maximum" fields for these so that the form validates those inputs for you.

You can take a look at https://rjsf-team.github.io/react-jsonschema-form/ and the "Numbers" example on there to see what I mention here done. 

Same comment applies to the other instruments as well, of course.

Reply:
@kmshin1397 Have you had experience making these sliders work with non-integers? I get weird things like: jsonschema.exceptions.ValidationError: 1.4 is not a multiple of 0.1.

Reply:
@mcoughlin Ah not outside just trying it briefly with that playground I linked when I was writing my review comment. It seemed to work fine on there as you can see [here](https://rjsf-team.github.io/react-jsonschema-form/#eyJmb3JtRGF0YSI6eyJudW1iZXIiOjMuMTQsImludGVnZXIiOjQyLCJudW1iZXJFbnVtIjoyLCJpbnRlZ2VyUmFuZ2UiOjIuMSwiaW50ZWdlclJhbmdlU3RlcHMiOjgwLCJudW1iZXJSYW5nZSI6Mi4xfSwic2NoZW1hIjp7InR5cGUiOiJvYmplY3QiLCJ0aXRsZSI6Ik51bWJlciBmaWVsZHMgJiB3aWRnZXRzIiwicHJvcGVydGllcyI6eyJudW1iZXJSYW5nZSI6eyJ0aXRsZSI6Ik51bWJlciByYW5nZSIsInR5cGUiOiJudW1iZXIiLCJtaW5pbXVtIjoxLCJtYXhpbXVtIjozLCJtdWx0aXBsZU9mIjowLjF9fX0sInVpU2NoZW1hIjp7Im51bWJlclJhbmdlIjp7InVpOndpZGdldCI6InJhbmdlIn19LCJ0aGVtZSI6ImRlZmF1bHQiLCJsaXZlU2V0dGluZ3MiOnsidmFsaWRhdGUiOmZhbHNlLCJkaXNhYmxlIjpmYWxzZSwib21pdEV4dHJhRGF0YSI6ZmFsc2UsImxpdmVPbWl0IjpmYWxzZX19). Not sure why you would get such an error.

Reply:
Maybe this is relevant? https://github.com/Julian/jsonschema/issues/247. Ideas for how to deal with it easily here? to be sure we get the proper type?

Reply:
It does seem to be related but it also seems like this issue was supposedly fixed a few years ago... https://github.com/rjsf-team/react-jsonschema-form/issues/342

I will look into this more.

Reply:
Thanks @kmshin1397. I currently have it commented out on my branch.

Reply:
@mcoughlin Okay, I'm honestly not sure why it's not working - this has definitely been solved and I even found the place in the rjsf code where the validation precision is set so this shouldn't happen (https://github.com/rjsf-team/react-jsonschema-form/blob/66623b9abad487c00a750014f50ff4d25248759d/packages/core/src/validate.js#L15). According to the [ajv documentation](https://github.com/ajv-validator/ajv/blob/master/docs/api.md#options) (the validator rsjf uses), given the `multipleOfPrecision=8` that rsjf uses, 1.9 should definitely pass for `multipleOf = 0.1`... It's not like we're using an old version of rjsf either. 

Oh well, I think for now, let's just revert this back to the default number input instead of the slider and get rid of the `multipleOf` property. We don't need it as a validation property and only were using it for the slider step. Should still set the `minimum` and `maximum` validators though.

Reply:
Ah darn. Ok so this is what the current version does if you want to look. You can also see what I was doing based on what is commented out... maybe you see something obvious?

Reply:
Yeah I think what you have now is fine (once you delete the old `"ui:widget": "range"` stuff not used anymore). I don't think this was anything wrong with what you were doing - I also tried getting it to work locally and was seeing the same errors you were. I'm not really sure what's going on but I think the slider is more of a nice-to-have anyways so this is fine in my opinion.

Reply:
Makes sense. Thanks for all the effort here.